File: leet-code-solutions/leetcode/minimum-height-trees.py
class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        numberOfEdges={}
        graph=defaultdict(list)
        for u,v in edges:
            graph[u].append(v)
            graph[v].append(u)
        for node in range(n):
            numberOfEdges[node]=len(graph[node])
        queue=deque()
        for node in range(n):
            if numberOfEdges[node]==1:
                queue.append(node)
        while queue:
            if n<=2:
                return list(queue)
            length=len(queue)
            for _ in range(length):
                cur=queue.popleft()
                n-=1
                for neighbor in graph[cur]:
                    numberOfEdges[neighbor]-=1
                    if numberOfEdges[neighbor]==1:
                        queue.append(neighbor)
        return [0]

        

File: leet-code-solutions/minimum-height-trees.go
func findMinHeightTrees(n int, edges [][]int) []int {

    //create graph and degree...for information storage
    graph:=make(map[int][]int)
    degree:=make([]int,n)

    //populate both storage unit...graph and degree
    for _,edge:= range edges{
        graph[edge[0]]=append(graph[edge[0]],edge[1])
        graph[edge[1]]=append(graph[edge[1]],edge[0])
        degree[edge[0]]+=1
        degree[edge[1]]+=1
    }

    //create queue and init with degree[node]==1
    queue:=[]int{}
    for node:=0;node<n;node++{
        if degree[node]==1{
            queue=append(queue,node)
        }
    }

    //perform bfs......until two or less nodes left inside the queue
    for len(queue)>0{
        length:=len(queue)
        if n<=2{
            return queue
        }
        for length>0{
            cur:=queue[0]
            queue=queue[1:]
            length--
            n--
            for _,neighbor:=range graph[cur]{
                degree[neighbor]-=1
                if degree[neighbor]==1{
                    queue=append(queue,neighbor)
                }
            }
        }

    }
    return []int{0}
    
}